plugins/cilium-cni/cni-install.sh

```diff
@@ -83,6 +84,14 @@ if [ "${CILIUM_CUSTOM_CNI_CONF}" = "true" ]; then
 	exit 0
 fi
 
+# Remove any active Cilium CNI configurations left over from previous installs
+# to make sure the one we're installing later will take effect.
+echo "Removing active Cilium CNI configurations from ${CNI_CONF_DIR}})..."
```

Comment:
Between this line and line 205 (`mv "${CNI_CONF_NAME}" "${CILIUM_CNI_CONF}"`) kubelet will be unable to create a pod. I know it will retry but ideally we should do `mv "${CNI_CONF_NAME}" "${CILIUM_CNI_CONF}"` first and then perform the deletes of the all confs that are not `${CNI_CONF_NAME}`

Reply:
This was done semi-intentionally because:
- this time window should be minimal
- this is during agent prestart, so not sure what Cilium CNI's pod creation behaviour is when the agent is not running yet
- there might be a timeframe where there are multiple Cilium CNI configs in the directory during upgrades, e.g. `05-cilium-cni.conflist` and a newer `05-cilium.conflist`.

I don't see the points above as problematic, it just seems safer to me to put the CNI directory into a clean state and only then install a new config. It also slightly simplifies the code.

Reply:
Answered inlie

> This was done semi-intentionally because:
> 
> * this time window should be minimal
yeah I know, but it can happen.

> * this is during agent prestart, so not sure what Cilium CNI's pod creation behaviour is when the agent is not running yet

Cilium CNI is smart enough and it will retry to connect to the agent multiple times before returns a failure to kubelet.

> * there might be a timeframe where there are multiple Cilium CNI configs in the directory during upgrades, e.g. `05-cilium-cni.conflist` and a newer `05-cilium.conflist`.
> 

IMO that is fine because in both cases Cilium is the one managing the network for the pod. Even if the 'wrong' CNI configuration is picked, that configuration is from Cilium.

> I don't see the points above as problematic, it just seems safer to me to put the CNI directory into a clean state and only then install a new config. It also slightly simplifies the code.

My main concern is during Cilium agent restarts, if the cilium pod is restarted, there will be a point in time where kubelet will fail hard to create pods because there isn't a single CNI configuration available.

Reply:
> Even if the 'wrong' CNI configuration is picked, that configuration is from Cilium.

True, but one might be a CNI chaining config, and the other might be a native IPAM config.

> if the cilium pod is restarted, there will be a point in time where kubelet will fail hard to create pods because there isn't a single CNI configuration available

IMO failing hard early is preferred over successfully launching a pod in the wrong networking config. The latter could mean  a security vulnerability (although it's a stretch).